### ISSUE TITLE
[pip-compile] Upgrading a dependency with `-P` that has extras fails with backtracking resolver

### ISSUE THREAD

**Issue by nwalsh1995** [jazzband/pip-tools#1806]
#### Environment Versions

1. Python version: 3.7.5
1. pip version: 23.0
1. pip-tools version: 6.12.2

#### Steps to replicate
1. Create `requirements.in` file. 

```
# requirements.in
googleapis-common-protos[grpc]==1.56.4
google-api-core[grpc]==1.32.0
```

2. Attempt upgrading  `googleapis-common-protos[grpc]` with `-P` and the backtracking resolver
```
pip-compile -v --resolver backtracking -P googleapis-common-protos\[grpc\] requirements.in
```
#### Expected result
`pip-compile` upgrades `googleapis-common-protos\[grpc\]` and extra packages. 

#### Actual result
```
pip._internal.exceptions.InstallationError: Constraints cannot have extras
```

Note: without `--resolver backtracking` the call succeeds after multiple rounds.
```
pip-compile -v -P googleapis-common-protos\[grpc\] requirements.in
```

Edit: the command also doesn't work with `--strip-extras`:
```
$ pip-compile -v --strip-extras --resolver=backtracking -P googleapis-common-protos\[grpc\] requirements.in

pip._internal.exceptions.InstallationError: Constraints cannot have extras
```

**Comment by pawelad** [comment]
You should be able to bypass that problem by passing extra `--strip-extras` option to `pip-tools`.

**Comment by nwalsh1995** [comment]
> You should be able to bypass that problem by passing extra `--strip-extras` option to `pip-tools`.

Hi @pawelad, it doesn't work with `--strip-extras` either:
```
$ pip-compile -v --strip-extras --resolver=backtracking -P googleapis-common-protos\[grpc\] requirements.in

pip._internal.exceptions.InstallationError: Constraints cannot have extras
```

**Comment by nwalsh1995** [comment]
FYI the patch from @deivid-rodriguez [here](https://github.com/dependabot/dependabot-core/pull/6551#issuecomment-1414378144) fixes this issue. Will see if Deivid wants to make the PR.

**Comment by deivid-rodriguez** [comment]
Happy to provide a PR but as I said over at the dependabot ticket, I did not manage to write a test for it.

**Comment by nwalsh1995** [comment]
This issue is fixed in #1808 🎉

**Comment by atugushev** [comment]
Completed via #1808.
